### PR TITLE
DRILL-7337: Add vararg UDFs support

### PR DESCRIPTION
**Pull request by vvysotskyi**
For details please see [DRILL-7337](https://issues.apache.org/jira/browse/DRILL-7337).

### CODE REVIEW

**Comment by cgivre** [comment]
This is a really useful feature!

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/DrillFuncHolder.java [hunk 1]
```diff
@@ -262,6 +260,78 @@ protected void addProtectedBlock(ClassGenerator<?> g, JBlock sub, String body, H
...
 
+  /**
+   * Checks whether current function is var arg one and declares array for storing vararg function arguments.
+   *
+   * @param model          code model to generate the code
+   * @param jBlock         block of code to be populated
+   * @param inputVariables array of input variables for current function
+   */
+  protected void declareVarArgArray(JCodeModel model, JBlock jBlock, HoldingContainer[] inputVariables) {
+    if (isVarArg()) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/DrillFuncHolder.java]: Not sure if this check should be done here, I would assume its caller responsibility first check `isVarArg()` and then call `declareVarArgArray`.

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/DrillFuncHolder.java [hunk 2]
```diff
@@ -298,11 +368,19 @@ public int getParamCount() {
   }
 
   public boolean isConstant(int i) {
-    return attributes.getParameters()[i].isConstant();
+    return getAttributeParameter(i).isConstant();
+  }
+
+  public ValueReference getAttributeParameter(int i) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/DrillFuncHolder.java]: Please add java doc thus there will be no need in the comment in the code.

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/FunctionConverter.java [hunk 3]
```diff
@@ -93,18 +96,30 @@
       // TODO(Julien): verify there are a few of those and we can load them
       Class<?> fieldClass = field.getFieldClass();
       if (param != null || output != null) {
+        if (Object[].class.isAssignableFrom(fieldClass)) {
+          fieldClass = fieldClass.getComponentType();
+          varArgsCount++;
+        } else if (varArgsCount > 0 && param != null) {
+          return failure("Vararg should be the last argument of function.", func, field);
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/FunctionConverter.java]:
```suggestion
          return failure("Vararg should be the last argument in the function.", func, field);
```

**Comment by arina-ielchiieva** [comment]
Also could you please show how varargs functions will be described in `sys.functions` table. Do we need new flag to filter them out?
Example:
```
apache drill (sys)> select * from sys.functions limit 2;
+------+---------------------------------+------------+----------+----------+
| name |            signature            | returnType |  source  | internal |
+------+---------------------------------+------------+----------+----------+
| !=   | BIGINT-REQUIRED,BIGINT-REQUIRED | BIT        | built-in | false    |
| !=   | BIT-REQUIRED,BIT-REQUIRED       | BIT        | built-in | false    |
+------+---------------------------------+------------+----------+----------+
```

**Comment by arina-ielchiieva** [comment]
+1, please squash the commits.

**Pull request merged**